File: sem2/projects/Firstclass/Firstclass/Fraction.cs
using System;

namespace FirstClass {
    class Fraction {
        private int numerator;
        private int denominator;
        private void Reduction() {
            if (numerator == 0) {
                denominator = 1;
            } else {
                int gcd = (int)GetGCD(denominator, numerator);
                numerator = numerator / gcd;
                denominator = denominator / gcd;
                if (denominator < 0) {
                    denominator = denominator * -1;
                    numerator = numerator * -1;
                }
            }
        }
        public static long GetGCD(long a, long b) {
            a = Math.Abs(a);
            b = Math.Abs(b);
            while (a != 0 && b != 0) {
                if (a > b) {
                    a = a % b;
                } else {
                    b = b % a;
                }
            }
            return a + b;
        }
        public Fraction(int num, int den) {
            if (den == 0) {
                throw new ArgumentException("Denominator can not be zero");
            }
            denominator = den;
            numerator = num;
            Reduction();
        }
        public int Denominator {
            get {
                return denominator;
            }
            set {
                if (value == 0) {
                    throw new ArgumentException("Denominator can not be zero");
                }
                denominator = value;
                Reduction();
            }
        }
        public int Numerator {
            get {
                return numerator;
            }
            set {
                numerator = value;
                Reduction();
            }
        }
        public static Fraction operator +(Fraction f1, Fraction f2) {
            long d = (long)f1.denominator * f2.denominator;
            long n = (long)f1.numerator * f2.denominator + (long)f1.denominator * f2.numerator;
            long gcd = GetGCD(d, n);
            d = d / gcd;
            n = n / gcd;
            return new Fraction((int)n, (int)d);
        }
        public static Fraction operator -(Fraction f) {
            Fraction f1 = new Fraction(f.numerator, f.denominator);
            f1.numerator = f1.numerator * -1;
            return f1;
        }
        public static Fraction operator -(Fraction f1, Fraction f2) {
            Fraction f = new Fraction(f2.numerator, f2.denominator);
            f = -f;
            return f1 + f;
        }
        public static Fraction operator /(Fraction f1, Fraction f2) {
            long n = (long)f1.numerator * f2.denominator;
            long d = (long)f1.denominator * f2.numerator;
            if (d == 0) {
                throw new ArgumentException("Denominator can not be zero");
            }
            long gcd = GetGCD(d, n);
            d = d / gcd;
            n = n / gcd;
            return new Fraction((int)n, (int)d);
        }
        public static Fraction operator *(Fraction f1, Fraction f2) {
            long n = (long)f1.numerator * f2.numerator;
            long d = (long)f1.denominator * f2.denominator;
            long gcd = GetGCD(d, n);
            d = d / gcd;
            n = n / gcd;
            return new Fraction((int)n, (int)d);
        }

        public static bool operator >(Fraction f1, Fraction f2) {
            Fraction f = f1 - f2;
            if (f.numerator > 0) {
                return true;
            } else {
                return false;
            }
        }
        public static bool operator ==(Fraction f1, Fraction f2) {
            Fraction f = f1 - f2;
            if (f.numerator == 0) {
                return true;
            } else {
                return false;
            }
        }
        public static bool operator !=(Fraction f1, Fraction f2) {
            if (!(f1 == f2)) {
                return true;
            } else {
                return false;
            }
        }
        public static bool operator <=(Fraction f1, Fraction f2) {
            if (!(f1 > f2)) {
                return true;
            } else {
                return false;
            }
        }
        public static bool operator >=(Fraction f1, Fraction f2) {
            if ((f1 > f2) || (f1 == f2)) {
                return true;
            } else {
                return false;
            }
        }
        public static bool operator <(Fraction f1, Fraction f2) {
            if (!(f1 >= f2)) {
                return true;
            } else {
                return false;
            }
        }
        public static Fraction operator +(Fraction f1, int a) {
            Fraction f2 = new Fraction(a, 1);
            return f1 + f2;
        }
        public static Fraction operator +(int a, Fraction f1) {
            Fraction f2 = new Fraction(a, 1);
            return f1 + f2;
        }
        public static Fraction operator -(Fraction f1, int a) {
            Fraction f2 = new Fraction(a, 1);
            return f1 - f2;
        }
        public static Fraction operator -(int a, Fraction f1) {
            Fraction f2 = new Fraction(a, 1);
            return f2 - f1;
        }
        public static Fraction operator *(Fraction f1, int a) {
            Fraction f2 = new Fraction(a, 1);
            return f1 * f2;
        }
        public static Fraction operator *(int a, Fraction f1) {
            Fraction f2 = new Fraction(a, 1);
            return f2 * f1;
        }
        public static Fraction operator /(Fraction f1, int a) {
            Fraction f2 = new Fraction(a, 1);
            return f1 / f2;
        }
        public static Fraction operator /(int a, Fraction f1) {
            Fraction f2 = new Fraction(a, 1);
            return f2 / f1;
        }
        public static bool operator >(Fraction f1, int a) {
            Fraction f2 = new Fraction(a, 1);
            return f1 > f2;
        }
        public static bool operator <(Fraction f1, int a) {
            Fraction f2 = new Fraction(a, 1);
            return f1 < f2;
        }
        public static bool operator ==(Fraction f1, int a) {
            Fraction f2 = new Fraction(a, 1);
            return f1 == f2;
        }
        public static bool operator >=(Fraction f1, int a) => !(f1 < a);
        public static bool operator <=(Fraction f1, int a) => !(f1 > a);
        public static bool operator !=(Fraction f1, int a) => !(f1 == a);
        public static bool operator >(int a, Fraction f1) {
            Fraction f2 = new Fraction(a, 1);
            return f2 > f1;
        }
        public static bool operator <(int a, Fraction f1) {
            Fraction f2 = new Fraction(a, 1);
            return f2 < f1;
        }
        public static bool operator ==(int a, Fraction f1) {
            Fraction f2 = new Fraction(a, 1);
            return f1 == f2;
        }
        public static bool operator >=(int a, Fraction f1) => !(a < f1);

        public static bool operator <=(int a, Fraction f1) => !(a > f1);
        public static bool operator !=(int a, Fraction f1) => !(f1 == a);
        public static explicit operator double(Fraction f) => (double)f.numerator / (double)f.denominator;
        public override bool Equals(object obj) {
            if (obj == null) {
                throw new NullReferenceException();
            }
            if (!(obj is Fraction)) {
                throw new ArgumentException("Argument should be Fraction type");
            }
            return (numerator * (obj as Fraction).denominator == (obj as Fraction).numerator * denominator);
        }
        public override int GetHashCode() => checked(numerator - denominator);
        public override string ToString() => String.Format($"{ numerator}/{ denominator}");
    }
}

File: sem1/projects C++/My class fraction/My class fraction/Fraction.cpp
#include "Fraction.h"
#include <cmath>

void Fraction::Reduction()
{
	if (numerator == 0) {
		denominator = 1;
	}
	else {
		long den1 = std::abs(denominator);
		long num1 = std::abs(numerator);
		while (den1 != 0 && num1 != 0) {
			if (den1 > num1) {
				den1 = den1 % num1;
			}
			else {
				num1 = num1 % den1;
			}
		}
		numerator /= den1 + num1;
		denominator /= den1 + num1;
		if (denominator < 0) {
			numerator *= -1;
			denominator *= -1;
		}
	}
}

Fraction::Fraction(long num, long den)
{
	if (den == 0) {
		throw std::runtime_error("denominator can not be zero");
	}
	numerator = num;
	denominator = den;
	Reduction();

}

long Fraction::GetNumerator() const
{
	return numerator;
}

long Fraction::GetDenominator() const
{
	return denominator;
}

void Fraction::SetNumerator(long num)
{
	numerator = num;
	Reduction();
}

void Fraction::SetDenominator(long den)
{
	denominator = den;
	Reduction();
	if (den == 0) {
		throw std::runtime_error("denominator can not be zero");
	}

}

Fraction Fraction::operator +(const Fraction& f) const
{
	long num = this->numerator * f.denominator + this->denominator * f.numerator;
    long den = this->denominator * f.denominator;
	return Fraction(num, den);
}

Fraction Fraction::operator -(const Fraction& f) const
{
	long num = this->numerator * f.denominator - this->denominator * f.numerator;
	long den = this->denominator * f.denominator;
	return Fraction(num, den);
}

Fraction Fraction::operator *(const Fraction& f) const
{
	long num = this->numerator * f.numerator;
	long den = this->denominator * f.denominator;
	return Fraction(num, den);
}

Fraction Fraction::operator /(const Fraction& f) const
{
	long num = this->numerator * f.denominator;
	long den = this->denominator * f.numerator;
	return Fraction(num, den);
}

Fraction Fraction::operator -() const
{
	return Fraction(-numerator, denominator);
}

bool Fraction::operator >(const Fraction& f) const
{
	if (this->numerator * f.denominator > f.numerator * this->denominator) {
		return true;
	}
	else {
		return false;
	}
}

bool Fraction::operator <(const Fraction& f) const
{
	if (this->numerator * f.denominator < f.numerator * this->denominator) {
		return true;
	}
	else {
		return false;
	}
}

bool Fraction::operator >=(const Fraction& f) const
{
	if (*this > f || *this == f) {
		return true;
	}
	else {
		return false;
	}
}

bool Fraction::operator <=(const Fraction& f) const
{
	if (*this == f || *this < f) {
		return true;
	}
	else {
		return false;
	}
}

bool Fraction::operator ==(const Fraction& f) const
{
	if (this->numerator * f.denominator == f.numerator * this->denominator) {
		return true;
	}
	else {
		return false;
	}
}

bool Fraction::operator!=(const Fraction& f) const
{
	if (!(*this == f)) {
		return true;
	}
	else {
		return false;
	}
}

double Fraction::GetDouble()
{
	return(static_cast<double>(numerator) / denominator);
}

Fraction operator +(const int& a, const Fraction& f2)
{
	Fraction f1(a, 1);
	return f1 + (f2);
}

Fraction operator -(const int& a, const Fraction& f2)
{
	Fraction f1(a, 1);
	return f1 - (f2);
}

Fraction operator *(const int& a, const Fraction& f2)
{
	Fraction f1(a, 1);
	return f1 * (f2);
}

Fraction operator /(const int& a, const Fraction& f2)
{
	Fraction f1(a, 1);
	return f1 / (f2);
}

Fraction operator +(const Fraction& f1, const int& a)
{
	Fraction f2(a, 1);
	return f1 + (f2);
}

Fraction operator -(const Fraction& f1, const int& a)
{
	Fraction f2(a, 1);
	return f1 - (f2);
}

Fraction operator *(const Fraction& f1, const int& a)
{
	Fraction f2(a, 1);
	return f1 * (f2);
}

Fraction operator /(const Fraction& f1, const int& a)
{
	Fraction f2(a, 1);
	return f1 * (f2);
}

bool operator >(const int& a, const Fraction& f2)
{
	Fraction f1(a, 1);
	if (f1.numerator * f2.denominator > f1.denominator * f2.numerator) {
		return true;
	}
	else {
		return false;
	}
}

bool operator <(const int& a, const Fraction& f2)
{
	Fraction f1(a, 1);
	if (f1.numerator * f2.denominator < f1.denominator * f2.numerator) {
		return true;
	}
	else {
		return false;
	}
}

bool operator >=(const int& a, const Fraction& f2)
{
	Fraction f1(a, 1);
	if (f1.numerator * f2.denominator >= f1.denominator * f2.numerator) {
		return true;
	}
	else {
		return false;
	}
}

bool operator <=(const int& a, const Fraction& f2)
{
	Fraction f1(a, 1);
	if (f1.numerator * f2.denominator <= f1.denominator * f2.numerator) {
		return true;
	}
	else {
		return false;
	}
}

bool operator !=(const int& a, const Fraction& f2)
{
	Fraction f1(a, 1);
	if (f1.numerator * f2.denominator != f1.denominator * f2.numerator) {
		return true;
	}
	else {
		return false;
	}
}

bool operator ==(const int& a, const Fraction& f2)
{
	Fraction f1(a, 1);
	if (f1.numerator * f2.denominator == f1.denominator * f2.numerator) {
		return true;
	}
	else {
		return false;
	}
}

bool operator >(const Fraction& f1, const int& a)
{
	Fraction f2(a, 1);
	if (f1.numerator * f2.denominator > f1.denominator * f2.numerator) {
		return true;
	}
	else {
		return false;
	}
}

bool operator <(const Fraction& f1, const int& a)
{
	Fraction f2(a, 1);
	if (f1.numerator * f2.denominator < f1.denominator * f2.numerator) {
		return true;
	}
	else {
		return false;
	}
}

bool operator >=(const Fraction& f1, const int& a)
{
	Fraction f2(a, 1);
	if (f1.numerator * f2.denominator >= f1.denominator * f2.numerator) {
		return true;
	}
	else {
		return false;
	}
}

bool operator <=(const Fraction& f1, const int& a)
{
	Fraction f2(a, 1);
	if (f1.numerator * f2.denominator <= f1.denominator * f2.numerator) {
		return true;
	}
	else {
		return false;
	}
}

bool operator !=(const Fraction& f1, const int& a)
{
	Fraction f2(a, 1);
	if (f1.numerator * f2.denominator != f1.denominator * f2.numerator) {
		return true;
	}
	else {
		return false;
	}
}

bool operator ==(const Fraction& f1, const int& a)
{
	Fraction f2(a, 1);
	if (f1.numerator * f2.denominator > f1.denominator * f2.numerator) {
		return true;
	}
	else {
		return false;
	}
}